### PR TITLE
[WIP] added backend directory and sql files

### PR DESCRIPTION
**Pull request by jrich8573**
Added application backend directory and create table `sql` files. Continuing to work through the sql create table code.

### CODE REVIEW

**Comment by jrich8573** [comment]
TODO

- [ ] add last three sql files
- [ ] added `fk` and reference declarations within the `sql` code


Before merging the PR:
- [ ] test against dev env database
- [ ] push tables to production database

**Comment by jrich8573** [comment]
Added all create table `sql` files.

File: backend/sql/create_grants_entity.sql [hunk 1]
```diff
@@ -0,0 +1,6 @@
+create table tbl_fact_granting_entity(
+            id smallint UNSIGNED
+           ,name varchar(50)
+           ,type char(8) check (type in ('external', 'internal'))
```
Comment (lamalex) [inline backend/sql/create_grants_entity.sql:4]: Inclined to say that strings like this should at least be ALL_CAPS like you have in UG/MS/PHP and ADMIN/FACULTY to differentiate them from arbitrary strings.

Even more inclined to make them integer values and then define an enum at the application level

**Comment by lamalex** [comment]
How do you intend these to be run? Manually?

File: backend/sql/create_grants_entity.sql (hunk 1 above)
Comment (jrich8573) [reply backend/sql/create_grants_entity.sql:4]: Internal/external should be all caps.

Comment (jrich8573) [reply backend/sql/create_grants_entity.sql:4]: You can certainly enum them at the db level too. Either way works for me.

Comment (jrich8573) [reply backend/sql/create_grants_entity.sql]: @lamalex locally I will create them manually, however, I think we need to have a more in-depth architecture conversation around the database. I am not opposed to scripting (which I am doing locally within my db container) the production creation. 

I am a little confused to your follow up questions. I would assume we will use GH for change management. I will add the drop table functional to the scripts later today, which to allow up to check for the existence of a not null table and, and if not null, we can drop and re create. 

One note, the create table code should only run on demand, not as a stage (or step) of the CI/CD pipeline.

File: backend/scripts/sql_run.sh [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+		export DB_STATUS
+		echo "`date` :Status: ${DB_STATUS}. Able to Connect..."
+	else
+		DB_STATUS="DOWN"
+		export DB_STATUS
+		echo "`date` :Status: DOWN . Not able to Connect."
+		echo "`date` :Not able to connect to database with Username:
+		"${user}" HostName: ""${host}" DB Port: "${port}"." SID: "${service}"."
+		echo "`date` :Exiting Script Run..."
+		exit
```
Comment (lamalex) [inline backend/scripts/sql_run.sh:49]: I think we should set a non-zero exit code here. Server being down seem like a condition we’d want to consider a failure case.

File: backend/scripts/sql_run.sh [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	then
+		for file in `dir -d $master`;
+		do
+		#for file in `cat extrasqlslist.txt` ;do
+		   echo "`date`:Executing file $file...";
+		   echo "`date`:__________________________________________";
+		   echo "`date`:SQL OUTPUT:";
+		   echo "`date`:__________________________________________";
+		   mysql -s ${user}/${password}@${host}/${service}<<-EOF
+		   @$file;
```
Comment (lamalex) [inline backend/scripts/sql_run.sh:78]: You’re running these interactively in the mysql repl right? Is there a reason you’re not using stdin? 
`mysql -s ${user}/${password}@${host}/${service} < <filename.sql>`

File: backend/sql/create_grants.sql [hunk 4]
```diff
@@ -0,0 +1,13 @@
+create table tbl_fact_grants(
```
Comment (lamalex) [inline backend/sql/create_grants.sql:1]: I think in my PR #2 you asked about checking if the tables exist. You can do `CREATE IF NOT EXISTS` which will only execute your `create` statement if the tables haven’t been created.

File: backend/scripts/sql_run.sh (hunk 3 above)
Comment (jrich8573) [reply backend/scripts/sql_run.sh:78]: @lamalex I really just want a way to auto the table build (and any other sql related data structure). We could run these at the shell, but I would prefer auto builds as we promote into prod.

File: backend/scripts/sql_run.sh (hunk 2 above)
Comment (jrich8573) [reply backend/scripts/sql_run.sh:49]: @lamalex yup. This is untested at this point. Just putting thoughts down first, and I will refine within my env before merging into master

File: backend/sql/create_grants.sql (hunk 4 above)
Comment (jrich8573) [reply backend/sql/create_grants.sql:1]: @lamalex that kind of what I thought too, just making sure we are on the same page.

**Comment by lamalex** [comment]
Yeah we can totally do that. I have duty tonight but I can pair on it tomorrow with you if you want

— Alex Launi

**Comment by jrich8573** [comment]
@lamalex no worries. Tomorrow evening works for me. Be safe.

**Comment by jrich8573** [comment]
Removed all `FK` references, will need to run the following after migration:

```sql
ALTER TABLE <<table_name>> ADD FOREIGN KEY <<key>> REFERENCES <<table(key)>>;
```

Tested against `mysql` latest container, by creating all tables manually. Did not test using the `sql_run.sh` script.

File: db/sql/006_create_purchases.sql [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+       ,grant_number smallint UNSIGNED
+       ,pname varchar(50)
+       ,qty int UNSIGNED
+       ,link varchar(100)
+       ,unit_cost float(7,2)
+       ,total_cost float(7,2)
+       ,purchaser smallint UNSIGNED
+       ,order_date datetime
+       ,deliver_date datetime
+       ,cancel_date datetime
```
Comment (lamalex) [inline db/sql/006_create_purchases.sql:12]: Same as above, recommend TimeStamp

File: db/sql/008_create_grants.sql [hunk 6]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE IF NOT EXISTS tbl_fact_grants(
+        id smallint UNSIGNED NOT NULL auto_increment PRIMARY KEY
+       ,grant_number varchar(20)
+       ,title varchar(30)
+       ,source_id smallint UNSIGNED
+       ,original_amt float(7,0)
+       ,balance float(7,2)
+       ,active varchar(7) check (active in ('PENDING','APPROVED', 'DENIED'))
```
Comment (lamalex) [inline db/sql/008_create_grants.sql]: ENUM is a safer and less verbose DT for list of string values

**Review by lamalex — CHANGES_REQUESTED** [review]

File: db/sql/003_create_users.sql [hunk 7]
```diff
@@ -0,0 +1,11 @@
+CREATE TABLE IF NOT EXISTS tbl_fact_users(
+       id int UNSIGNED AUTO_INCREMENT NOT NULL,
+       name varchar(30) NOT NULL,
+       email varchar(15) NOT NULL unique, 
+       password varchar(100) NOT NULL,
+       created_date Timestamp,
+       update_date TimeStamp,
+       user_role ENUM("FACULTY", "ADMINISTRATOR"),
+       department int UNSIGNED NOT NULL,
```
Comment (lamalex) [inline db/sql/003_create_users.sql:9]: shouldn't this be a foreign key?

File: db/sql/004_create_students.sql [hunk 8]
```diff
@@ -0,0 +1,6 @@
+CREATE TABLE IF NOT EXISTS tbl_fact_student(
+        uin smallint UNSIGNED NOT NULL auto_increment PRIMARY KEY,
+        sname varchar(50),
+        department smallint UNSIGNED,
```
Comment (lamalex) [inline db/sql/004_create_students.sql:4]: same as above, foreign key?

**Review by lamalex — CHANGES_REQUESTED** [review]
There are no foreign keys declared. Was the intentional?
Additionally I'm hesitant to add a bunch of tables before we need them. I'd rather add them as we build, but I'm not married to either way

**Comment by jrich8573** [comment]
- [ ] need to edit grants table make grant_number an unsigned int 
- [ ] need to add  FK references

**Comment by lamalex** [comment]
Closing due to merging #12 which superceded this one

**Pull request closed without merging**